### PR TITLE
Topic/cq update

### LINKED ISSUE #115: Implement generic queue structure that can be used for both CQ and EQ.

### PR DESCRIPTION
**Pull request by bturrubiates**
Closes #115. Should be merged after #121.

### CODE REVIEW

File: prov/gni/src/gnix_cq.c [hunk 1]
```diff
@@ -427,17 +360,41 @@ static ssize_t gnix_cq_readfrom(struct fid_cq *cq, void *buf, size_t count,
...
+	if (!_gnix_queue_peek(cq_priv->events))
+		return -FI_EAGAIN;
+
+	fastlock_acquire(&cq_priv->lock);
+
+	while (count--) {
+		temp = _gnix_queue_dequeue(cq_priv->events);
+		event = container_of(temp, struct gnix_cq_entry, item);
+
+		assert(buf);
```
Comment (hppritcha) [inline prov/gni/src/gnix_cq.c]: maybe move the assert(buf) to above the lock?

File: prov/gni/src/gnix_cq.c [hunk 2]
```diff
@@ -345,23 +289,20 @@ ssize_t _gnix_cq_add_error(struct gnix_fid_cq *cq, void *op_context,
...
-				     struct gnix_cq_entry, item);
-	} else {
-		event = alloc_cq_entry(sizeof(struct fi_cq_err_entry));
-	}
-
-	if (!event) {
+	item = _gnix_queue_get_free(cq->errors);
+	if (!item) {
 		GNIX_ERR(FI_LOG_CQ, "error creating error entry\n");
 		return -FI_ENOMEM;
```
Comment (hppritcha) [inline prov/gni/src/gnix_cq.c]: need to release lock here?

File: prov/gni/src/gnix_cq.c [hunk 3]
```diff
@@ -345,23 +289,20 @@ ssize_t _gnix_cq_add_error(struct gnix_fid_cq *cq, void *op_context,
...
-	if (!slist_empty(&cq->err_free)) {
-		event = container_of(slist_remove_head(&cq->ev_free),
-				     struct gnix_cq_entry, item);
-	} else {
-		event = alloc_cq_entry(sizeof(struct fi_cq_err_entry));
-	}
-
-	if (!event) {
+	item = _gnix_queue_get_free(cq->errors);
+	if (!item) {
```
Comment (hppritcha) [inline prov/gni/src/gnix_cq.c:301]: need to release lock before returning

**Comment by hppritcha** [comment]
found a few things that need fixing, esp. the lock, but otherwise looks good.

File: prov/gni/src/gnix_cq.c (hunk 3 above)
Comment (bturrubiates) [inline prov/gni/src/gnix_cq.c:301]: Good point. I just realized that was actually in some of the extraneous code I tried to phase out. By that I meant the one that was going to be in it's own separate PR of minor fixes.

File: prov/gni/src/gnix_cq.c (hunk 1 above)
Comment (bturrubiates) [inline prov/gni/src/gnix_cq.c]: Yeah, good point.

**Comment by bturrubiates** [comment]
@hppritcha Updated. Please review again.

**Comment by hppritcha** [comment]
:+1:

**Pull request merged**